File: platform_channel_events/android/app/src/main/kotlin/com/raywenderlich/platform_channel_events/ImageStreamHandler.kt
package com.raywenderlich.platform_channel_events

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.*
import java.io.ByteArrayOutputStream

class ImageStreamHandler(private var activity: Activity?) : EventChannel.StreamHandler {
    private var eventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        val quality = (arguments as Map<*, *>)["quality"] as Double
        val chunkSize = arguments["chunkSize"] as Int
        dispatchImageEvents(quality, chunkSize)
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        activity = null
    }

    private fun dispatchImageEvents(quality: Double, chunkSize: Int) {
        GlobalScope.launch(Dispatchers.IO) {
            if (activity == null) return@launch

            // Decode the drawable
            val bitmap = BitmapFactory.decodeResource(activity!!.resources, R.drawable.cute_cat_unsplash)

            // Compress the drawable using the quality passed from Flutter
            val stream = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, (quality * 100).toInt(), stream)

            // Convert the compressed image stream to byte array
            val byteArray = stream.toByteArray()

            // Dispatch the first event (which is the size of the array/image)
            withContext(Dispatchers.Main) {
                eventSink?.success(byteArray.size)
            }

            // Split the array into chunks using the chunkSize passed from Flutter
            val parts = byteArray.size / chunkSize
            val chunks = byteArray.toList().chunked(parts)

            // Loop through the chunks and dispatch each chuck to Flutter
            chunks.forEach {
                // Mimic buffering with a 50 mills delay
                delay(50)
                withContext(Dispatchers.Main) {
                    eventSink?.success(it)
                }
            }
            withContext(Dispatchers.Main) {
                // Finally, dispatch an event to indicate the end of the image stream
                eventSink?.success(Constants.eof)
            }
        }
    }

}

File: platform_channel_events/ios/Runner/ImageStreamHandler.swift

import Foundation

class ImageStreamHandler: NSObject, FlutterStreamHandler {
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        let args = arguments as! Dictionary<String, Any>
        let quality = args["quality"] as! Double
        let chunkSize = args["chunkSize"] as! Int
        dispatchImageEvents(quality: quality, chunkSize: chunkSize, eventSink: events)
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
    
    func dispatchImageEvents(quality: Double, chunkSize: Int, eventSink: @escaping FlutterEventSink) -> Void {
        
        // Load the image into memory
        guard let image  = UIImage(named: "cute_dog_unsplash.jpg"),
              
        // Compress the image using the quality passed from Flutter
        let data = image.jpegData(compressionQuality: CGFloat(quality)) else {return}
        
        // Get the size of the image
        let length = data.count
        
        // Dispatch the first event (which is the size of the image)
        eventSink(length)
        
        DispatchQueue.global(qos: .background).async {
            
            // Split the image into chunks using the chunkSize passed from Flutter
            let parts = length / chunkSize
            var offset = 0
            
            // Repeat this block of statements until we have dispatched the last chunk
            repeat {
                // Mimic buffering with a 1 mill delay
                usleep(1000)
                
                let thisChunkSize = ((length - offset) > parts) ? parts : (length - offset)
                let chunk  = data.subdata(in: offset..<offset + thisChunkSize)
                
                // Dispatch each chunk to Flutter
                eventSink(chunk)
                offset += thisChunkSize
                
            } while (offset < length)
            
            // Dispatch an event to indicate the end of the stream
            eventSink(Constants.eof)
        }
        
    }

    
}
